File: lib/utils/note_search.dart
import 'package:lpinyin/lpinyin.dart';

/// Small, purpose-built search helper for notes.
///
/// Supports:
/// - Case-insensitive latin/number search
/// - Chinese pinyin full/initials search (via lpinyin)
/// - Simple scoring for sorting results
class NoteSearchIndex {
  final String sourceText;
  final String normalizedText;
  final String normalizedPinyin;
  final String normalizedInitials;

  NoteSearchIndex._({
    required this.sourceText,
    required this.normalizedText,
    required this.normalizedPinyin,
    required this.normalizedInitials,
  });

  factory NoteSearchIndex.fromText(String text) {
    final normalized = _normalize(text);
    return NoteSearchIndex._(
      sourceText: text,
      normalizedText: normalized,
      normalizedPinyin: _normalize(_safePinyin(text)),
      normalizedInitials: _normalize(_safeInitials(text)),
    );
  }
}

class NoteSearchMatch {
  final bool matched;
  final int score;

  const NoteSearchMatch._(this.matched, this.score);

  static const none = NoteSearchMatch._(false, 0);
  static const passthrough = NoteSearchMatch._(true, 100);
}

class NoteSearchMatcher {
  NoteSearchMatcher._();

  static NoteSearchMatch match(String query, NoteSearchIndex index) {
    final q = _normalize(query);
    if (q.isEmpty) return NoteSearchMatch.passthrough;

    final text = index.normalizedText;
    if (text == q) return const NoteSearchMatch._(true, 100);
    if (text.startsWith(q)) return const NoteSearchMatch._(true, 90);
    if (text.contains(q)) return const NoteSearchMatch._(true, 75);

    final pinyin = index.normalizedPinyin;
    if (pinyin.startsWith(q)) return const NoteSearchMatch._(true, 65);
    if (pinyin.contains(q)) return const NoteSearchMatch._(true, 55);

    final initials = index.normalizedInitials;
    if (initials.startsWith(q)) return const NoteSearchMatch._(true, 50);
    if (_isSubsequence(q, initials)) return const NoteSearchMatch._(true, 40);

    return NoteSearchMatch.none;
  }
}

final RegExp _split = RegExp(r'[^a-z0-9\u4e00-\u9fff]+', caseSensitive: false);

String _normalize(String input) => input.toLowerCase().replaceAll(_split, '');

String _safePinyin(String input) {
  if (input.trim().isEmpty) return '';
  try {
    return PinyinHelper.getPinyinE(input);
  } catch (_) {
    return '';
  }
}

String _safeInitials(String input) {
  if (input.trim().isEmpty) return '';
  try {
    return PinyinHelper.getShortPinyin(input);
  } catch (_) {
    return '';
  }
}

bool _isSubsequence(String needle, String haystack) {
  if (needle.isEmpty) return true;
  var i = 0;
  for (var j = 0; j < haystack.length && i < needle.length; j++) {
    if (haystack[j] == needle[i]) i++;
  }
  return i == needle.length;
}

File: src/lib/note-search.js
import { pinyin } from 'pinyin-pro';

const SPLIT_RE = /[^a-z0-9\u4e00-\u9fff]+/gi;

/** @param {string} s */
function normalize(s) {
  return String(s || '')
    .toLowerCase()
    .replace(SPLIT_RE, '');
}

/** @param {string} text */
function getPinyin(text) {
  try {
    // @ts-ignore
    return pinyin(text, { toneType: 'none', type: 'string' }).replace(/\s/g, '') || '';
  } catch {
    return '';
  }
}

/** @param {string} text */
function getInitials(text) {
  try {
    // @ts-ignore
    return pinyin(text, { pattern: 'first', toneType: 'none', type: 'string' }).replace(/\s/g, '') || '';
  } catch {
    return '';
  }
}

/** 
 * @param {string} needle
 * @param {string} haystack
 */
function isSubsequence(needle, haystack) {
  if (!needle) return true;
  let i = 0;
  for (let j = 0; j < haystack.length && i < needle.length; j++) {
    if (haystack[j] === needle[i]) i++;
  }
  return i === needle.length;
}

/**
 * Match query against note text. Supports:
 * - Case-insensitive text match
 * - Chinese pinyin full match
 * - Pinyin initials match
 * @param {string} query - Search query
 * @param {string} text - Note text
 * @returns {{ matched: boolean, score: number }}
 */
export function matchNote(query, text) {
  const q = normalize(query);
  if (!q) return { matched: true, score: 100 };

  const normText = normalize(text);
  if (normText === q) return { matched: true, score: 100 };
  if (normText.startsWith(q)) return { matched: true, score: 90 };
  if (normText.includes(q)) return { matched: true, score: 75 };

  const py = normalize(getPinyin(text));
  if (py.startsWith(q)) return { matched: true, score: 65 };
  if (py.includes(q)) return { matched: true, score: 55 };

  const initials = normalize(getInitials(text));
  if (initials.startsWith(q)) return { matched: true, score: 50 };
  if (isSubsequence(q, initials)) return { matched: true, score: 40 };

  return { matched: false, score: 0 };
}
